src/monitor/mod.rs

```diff
@@ -0,0 +1,570 @@
+                                }
+
+                                Node::new(
+                                    node_id.clone(),
+                                    "",
+                                    current_region.clone(),
+                                    NodeKind::StrictInput { output },
+                                )
+                            } else {
+                                self.current_scope = node_id.clone();
+                                self.region_stack.push(RegionId::root());
+                                Node::new(
+                                    node_id.clone(),
+                                    "",
+                                    current_region.clone(),
+                                    NodeKind::Circuit {
+                                        iterative: event.is_iterative_subcircuit_event(),
+                                        children: HashMap::new(),
+                                        region: Region::new(RegionId::root(), Cow::Borrowed("")),
+                                    },
+                                )
+                            };
+
+                        children.insert(local_node_id, new_node);
+                        region.get_region(&current_region).nodes.push(local_node_id);
+                        Ok(())
+                    }
+                    _ => Err(TraceError::NotACircuit(parent_id)),
+                }
+            } else {
+                if node_id != &self.current_scope {
+                    return Err(TraceError::UnexpectedNodeId);
+                }
+                self.current_scope = parent_id;
+                self.region_stack.pop();
+                Ok(())
+            }
+        } else if event.is_edge_event() {
+            let from = event.from().unwrap();
+            let to = event.to().unwrap();
+            let kind = event.edge_kind().unwrap();
+            self.circuit
+                .node_ref(from)
+                .ok_or_else(|| TraceError::UnknownNode(from.clone()))?;
+            self.circuit
+                .node_ref(to)
+                .ok_or_else(|| TraceError::UnknownNode(to.clone()))?;
+            self.circuit.add_edge(from, to, kind);
+            Ok(())
+        } else {
+            match event {
+                CircuitEvent::PushRegion { name } => {
+                    self.push_region(name.clone());
+                    Ok(())
+                }
+                CircuitEvent::PopRegion => self.pop_region(),
+                _ => panic!("unknown event"),
+            }
+        }
+    }
+
+    fn scheduler_event(&mut self, event: &SchedulerEvent) -> Result<(), TraceError> {
+        // eprintln!("scheduler event: {}", event);
+        if !self.running() {
+            return Err(TraceError::InvalidEvent(Cow::from(
+                "scheduler event received after circuit execution has terminated",
+            )));
+        }
+        let current_node_id = self.current_node_id();
```

Comment:
You use the current node in pretty much every arm, hoist the expression up to here and reuse it
```suggestion
        let current_node_id = self.current_node_id();
        let current_node = self.circuit.get_node(&current_node_id).unwrap();
```